### PR TITLE
adds dns registration at project init

### PR DESCRIPTION
**Pull request by andrewxhill**

### CODE REVIEW

File: cmd/textiled/main.go [hunk 1]
```diff
@@ -212,7 +241,14 @@ var rootCmd = &cobra.Command{
 		addrGatewayHost := cmd.AddrFromStr(configViper.GetString("addr.gateway.host"))
 		addrGatewayUrl := configViper.GetString("addr.gateway.url")
 
-		addrFilecoinApi := cmd.AddrFromStr(configViper.GetString("addr.filecoin.api"))
+		var addrFilecoinApi ma.Multiaddr
+		if filStr := configViper.GetString("addr.filecoin.api"); filStr != "" {
+			addrFilecoinApi = cmd.AddrFromStr(filStr)
+		}
+
```
Comment (andrewxhill) [inline cmd/textiled/main.go]: @asutula i needed to add this step so you could pass an empty address and not run api (as there is a nil check later that could never have been nil afai can tell)

File: collections/projects.go [hunk 2]
```diff
@@ -37,12 +43,31 @@ func (p *Projects) GetStoreID() *uuid.UUID {
 }
 
 func (p *Projects) Create(ctx context.Context, name, scope, fcWalletAddress string) (*Project, error) {
+	// Create subdomain for the project
+	domain := ""
+	records := []*dns.Record{}
+	if p.dnsManager.Started {
+		safesd, err := dns.CreateURLSafeSubdomain(name, 8)
```
Comment (andrewxhill) [inline collections/projects.go]: appends an 8 char string after the proj name for unique subdomains

File: collections/projects.go [hunk 3]
```diff
@@ -37,12 +43,31 @@ func (p *Projects) GetStoreID() *uuid.UUID {
...
+		if err != nil {
+			return nil, err
+		}
+		rds, err := p.dnsManager.NewDNSLink(safesd, defaultHash)
```
Comment (andrewxhill) [inline collections/projects.go]: went with DNS link to start. probably a good one to do all time. we can run the UpdateRecord method to change the dnslink has pointer preeeetty nicely. currently it defaults to the 404 page (as a hash), but we can create a similar style one for an empty project

File: collections/projects.go [hunk 4]
```diff
@@ -8,20 +8,26 @@ import (
...
+var defaultHash = "QmSNWjbDkafwxWUGWTyP1ko4J1CW2WeNfeoiCuTUW2YnDY"
+
 type Project struct {
 	ID              string
 	Name            string
 	Scope           string // user or team
 	StoreID         string
-	Created         int64
+	Domain          string
+	Records         []*dns.Record
```
Comment (andrewxhill) [inline collections/projects.go]: made this an array for max flexibility. DNSlinks for example require 2 records (one for the cname and one for the txt record)

File: cmd/textiled/main.go [hunk 5]
```diff
@@ -68,6 +69,22 @@ var (
...
+			Key:      "dns.zone_id",
+			DefValue: "",
+		},
+		"dnsToken": {
+			Key:      "dns.token",
+			DefValue: "",
+		},
+		"defaultProjectHash": {
+			Key:      "project_hash",
+			DefValue: "QmSNWjbDkafwxWUGWTyP1ko4J1CW2WeNfeoiCuTUW2YnDY",
```
Comment (sanderpick) [inline cmd/textiled/main.go]: What's the advantage of using a hash for project pages? It's gonna add a bunch of complexity. A project page will need a URL path off the gateway, which could just be mapped to the project subdomain.

File: collections/projects.go [hunk 6]
```diff
@@ -37,12 +42,31 @@ func (p *Projects) GetStoreID() *uuid.UUID {
 }
 
 func (p *Projects) Create(ctx context.Context, name, scope, fcWalletAddress string) (*Project, error) {
+	// Create subdomain for the project
+	domain := ""
```
Comment (sanderpick) [inline collections/projects.go]: var domain string

File: collections/projects.go [hunk 7]
```diff
@@ -37,12 +42,31 @@ func (p *Projects) GetStoreID() *uuid.UUID {
...
+	records := []*dns.Record{}
```
Comment (sanderpick) [inline collections/projects.go]: var records []*dns.Record

File: dns/dns.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+	"time"
+
+	"github.com/cloudflare/cloudflare-go"
+	logging "github.com/ipfs/go-log"
+	"github.com/textileio/go-threads/util"
+)
+
+var (
+	log         = logging.Logger("dns")
+	ipfsGateway = "www.cloudflare-ipfs.com" // future could be set to project's gateway
```
Comment (sanderpick) [inline dns/dns.go]: can be const

File: dns/dns.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+			return nil, err
+		}
+	}
+
+	client := &Manager{
+		domain: domain,
+		debug:  debug,
+	}
+
+	if token != "" {
```
Comment (sanderpick) [inline dns/dns.go]: Is there a reason to create the Manager if token is not present?

File: dns/dns.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+		client.api = api
+		client.zoneID = zoneID
+		client.Started = true
+	}
+	return client, nil
+}
+
+// NewCNAME enters a new dns record for a CNAME
+func (m *Manager) NewCNAME(subdomain string, target string) (*Record, error) {
+	// parse err here
```
Comment (sanderpick) [inline dns/dns.go]: left over comment?

File: dns/dns.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	}
+	return nil
+}
+
+// Delete removes a record by ID from dns
+func (m *Manager) Delete(recordID string) error {
+	return m.api.DeleteDNSRecord(m.zoneID, recordID)
+}
+
+//CreateDNSLinkName converts a subdomain into the Name format for dnslink TXT entries
```
Comment (sanderpick) [inline dns/dns.go]: Nit: spaces needed after // and a period at the end of the doc string.

File: dns/dns.go [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+//CreateDNSLinkContent converts a hash into the Content format for dnslink TXT entries
+func CreateDNSLinkContent(hash string) string {
+	return fmt.Sprintf("dnslink=/ipfs/%s", hash)
+}
+
+//CreateURLSafeSubdomain returns a url safe subdomain with optional suffix
+func CreateURLSafeSubdomain(subdomain string, suffix int) (string, error) {
+	sfx := ""
```
Comment (sanderpick) [inline dns/dns.go]: var sfx string

File: dns/dns.go [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+		var seededRand *rand.Rand = rand.New(
+			rand.NewSource(time.Now().UnixNano()))
+
+		b := make([]byte, suffix)
+		for i := range b {
+			b[i] = charset[seededRand.Intn(len(charset))]
+		}
+		sfx = fmt.Sprintf("-%s", string(b))
+	}
+	reg, err := regexp.Compile("[^a-zA-Z0-9]+")
```
Comment (sanderpick) [inline dns/dns.go]: This should be compiled in the package `init` function.

File: dns/dns.go [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+//CreateDNSLinkContent converts a hash into the Content format for dnslink TXT entries
+func CreateDNSLinkContent(hash string) string {
+	return fmt.Sprintf("dnslink=/ipfs/%s", hash)
+}
+
+//CreateURLSafeSubdomain returns a url safe subdomain with optional suffix
+func CreateURLSafeSubdomain(subdomain string, suffix int) (string, error) {
+	sfx := ""
+	if suffix > 0 {
+		charset := "abcdefghijklmnopqrstuvwxyz0123456789"
```
Comment (sanderpick) [inline dns/dns.go]: package constant?

File: dns/dns.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+		for i := range b {
+			b[i] = charset[seededRand.Intn(len(charset))]
+		}
+		sfx = fmt.Sprintf("-%s", string(b))
+	}
+	reg, err := regexp.Compile("[^a-zA-Z0-9]+")
+	if err != nil {
+		return "", nil
+	}
+	safestr := reg.ReplaceAllString(strings.ToLower(subdomain), "")
```
Comment (sanderpick) [inline dns/dns.go]: Is this needed if the charset is limited to letters and number?

File: cmd/textiled/main.go (hunk 1 above)
Comment (asutula) [reply cmd/textiled/main.go]: Nice. I should have done it like this anyway. Good call.

**Pull request closed without merging**